### DISCUSSION TITLE
A2DP -> MP3 Encoder -> USB Storage Class  :  (Bluetooth/USB mp3 dongle) with ESP32

### DISCUSSION THREAD

**Discussion by demon740** [pschatzmann/arduino-audio-tools#566] [Ideas]
there are some USB/Bluetooth dongles that you can connect to them as Bluetooth audio and the stream the audio as a mp3 file on the USB flash.
these devices are very simple. they are like a simple USB disk and you can connect them to all mp3 compatible players and make them as a Bluetooth player,  but all of them (that I have tested) have a big delay (about 20-30 sec) and many other bugs.

I know that some of ESP32's have USB Storage Class feature. 
I think is a great project to have all of these libs (A2DP, MP3 Encoder, USB Storage class) together.
thanks

**Comment by pschatzmann** [comment]
I don't see how the USB Storage Class would be of any value for audio...
So, I am not sure if I understand you correctly. Or do you mean the same like [this](https://github.com/pschatzmann/arduino-audio-tools/discussions/120) ?

**Comment by demon740** [comment]
very thanks for your replyImagine a USB flash that contains a mp3 file and you can play it for example with a car audio player.imagine that the content of the mp3 file is dynamically encoded from another source of audio (for example a Bluetooth connection or an aux input).this device is useful for adding audio Bluetooth feature to an audio player that just can play mp3 file (USB flash disk) and has not aux input.

I already have two of these devices but they have BIG delays and sometimes the sound is skipped or plays faster.I thought we can use these libraries on ESP32 module: Bluetooth AD2P + Mp3 Encoder + USB Mass Storage Class  
thanks for your guidance.

**Reply by pschatzmann** [reply]
Would you know if your car radio supports wav files ?
I expect with A2DP there is not enough PROGMEM and RAM to support any other decoder...

**Comment by demon740** [comment]
many of USB audio players, support PCM WAV file either.I think, it could be fine.

    On Thursday, January 12, 2023 at 08:12:46 PM GMT+3:30, Phil Schatzmann ***@***.***> wrote:  
 
 


Would you know if your car radio supports wav files ?
I expect with A2DP there is not enough PROGMEM and RAM to support any other decoder...

**Comment by Silenttttttt** [comment]
A bit late to this one (a couple years!), but I actually built it — confirmed working end-to-end on real hardware, including on my actual car radio (a Kenwood KDC-MP8090U).

One thing worth flagging for anyone else looking at this: it needs **two ESP32 boards**, not one. A "classic" ESP32 (Bluetooth Classic/A2DP) does the BT sink + real-time Shine MP3 encoding and sends the encoded bytes over a wired UART to an ESP32-S3, which does the actual USB-MSC device role. No single ESP32 chip does both halves — the classic chip has A2DP but no USB-OTG, and the S2/S3 chips have USB-OTG but only BLE, not Bluetooth Classic — so it ends up being a two-board bridge rather than a single-chip solution.

The USB-MSC side works by presenting a fixed-size FAT12 volume with one MP3 file — the file's declared size is fixed at mount time (real USB-MSC host drivers cache it and never re-check), but the actual bytes are served from a live ring buffer continuously overwritten with fresh audio. Same basic trick as the commercial dongles you mentioned, built specifically to avoid their delay/skip issues — confirmed clean, continuous audio on the real radio, roughly 10 seconds from power-up to hearing sound.

Repo: https://github.com/Silenttttttt/esp32-bt-usb-mp3-dongle — built with arduino-audio-tools (Shine encoder + A2DP sink) for the classic ESP32 side. Full build log and every bug found along the way is in progress/STATUS.md if it's useful to anyone else hitting the same wall.

**Comment by demon740** [comment]
Thank you very much for your significant effort and the clean results.

As you may know, the Espressif has many new series of ESP32 MCUs with a variety feature list.
it seems there are two simpler solution: 1- using a standalone Bluetooth chipset that supports A2DP with an ESP32 with OTG-USB Storage support (Like S2 or S3)
2- using new  ESP32-S31 (see this query from chatgpt: https://chatgpt.com/share/6aae00b6-36d4-83eb-aed5-d851aec74361 )

Although still your architecture will won because of using poplar and common boards(chipset).

Regards
    On Saturday, September 19, 2026 at 04:47:40 AM GMT+3:30, Muni Besen ***@***.***> wrote:  
 
 


A bit late to this one (a couple years!), but we actually built it — confirmed working end-to-end on real hardware, including on my actual car radio (a Kenwood KDC-MP8090U).

One thing worth flagging for anyone else looking at this: it needs two ESP32 boards, not one. A "classic" ESP32 (Bluetooth Classic/A2DP) does the BT sink + real-time Shine MP3 encoding and sends the encoded bytes over a wired UART to an ESP32-S3, which does the actual USB-MSC device role. No single ESP32 chip does both halves — the classic chip has A2DP but no USB-OTG, and the S2/S3 chips have USB-OTG but only BLE, not Bluetooth Classic — so it ends up being a two-board bridge rather than a single-chip solution.

The USB-MSC side works by presenting a fixed-size FAT12 volume with one MP3 file — the file's declared size is fixed at mount time (real USB-MSC host drivers cache it and never re-check), but the actual bytes are served from a live ring buffer continuously overwritten with fresh audio. Same basic trick as the commercial dongles you mentioned, built specifically to avoid their delay/skip issues — confirmed clean, continuous audio on the real radio, roughly 10 seconds from power-up to hearing sound.

Repo: https://github.com/Silenttttttt/esp32-bt-usb-mp3-dongle — built with arduino-audio-tools (Shine encoder + A2DP sink) for the classic ESP32 side. Full build log and every bug found along the way is in progress/STATUS.md if it's useful to anyone else hitting the same wall.

—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications, keep track of coding agent tasks and review pull requests on the go with GitHub Mobile for iOS and Android. Download it today!
